src/cpp/database/database.cpp

```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
     return entities;
 }
 
+void Database::erase(
+        EntityId& domain_id)
+{
+    // Check that the given domain_id corresponds to a known monitor.
+    // Upper layer ensures that the monitor has been stopped.
+    // Upper layer also ensures that the monitor_id is valid and corresponds to a known domain.
+    assert(EntityKind::DOMAIN == get_entity_kind(domain_id));
+
+    // The mutex should be taken only once. get_entity_kind already locks.
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    for (auto reader : datareaders_[domain_id])
+    {
+        // Unlink related locators
+        for (auto locator : reader.second.get()->locators)
+        {
+            locators_[locator.first].get()->data_readers.erase(reader.first);
+        }
+        // Clear entity
+        reader.second.get()->DDSEndpoint::clear();
+        reader.second.get()->clear();
+    }
+    // Clear datareaders map
+    datareaders_[domain_id].clear();
```

Comment:
Should remove the entry with key `domain_id` in the `datareaders_` map too, not just `clear` it. Same in the rest of entities

Reply:
The erase test has been completed to check that the maps are correctly erased in 460eaf2347095ad626848ba348254c32ef9e0cf9

The implementation has been changed in 13ff4e2fa03ae21d60a1fb12cb5dd20a00e0bf03